### PR TITLE
[POAE7-2741] create CiderNextgenBenchmarkBase framework and migrate FilterTest to nextgen test framework

### PR DESCRIPTION
**Pull request by jikunshang**
### What changes were proposed in this pull request?

create CiderNextgenBenchmarkBase framework and migrate FilterTest to nextgen test framework.
fix some issues of CiderNextgenTestBase. 
assertQueryIgnoreOrder is not ready yet, disable some related cases.

### Why are the changes needed?

for benchmark framework and jenkins integration.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

File: src/cider/tests/nextgen/functional/FilterTest.cpp [hunk 1]
```diff
@@ -220,123 +211,126 @@ TEST_F(CiderFilterSequenceTestNG, multiFilter) {
...
-  assertQueryArrowIgnoreOrder(
+  assertQueryIgnoreOrder(
       "SELECT col_1, col_2, col_3 FROM test WHERE col_2 >= col_3 or col_2 <= col_1", "");
-  assertQueryArrow("SELECT col_1, col_5 FROM test WHERE col_1 < col_5 AND col_5 > 0");
+  assertQuery("SELECT col_1, col_5 FROM test WHERE col_1 < col_5 AND col_5 > 0");
 }
 
 TEST_F(CiderFilterRandomTestNG, complexFilter) {
-  assertQueryArrowIgnoreOrder(
+  assertQueryIgnoreOrder(
```
Comment (yma11) [inline src/cider/tests/nextgen/functional/FilterTest.cpp:293]: Should this `assertQueryIgnoreOrder` also be skipped? Seems handle is not consistent.

File: src/cider/tests/utils/DuckDbQueryRunner.cpp [hunk 2]
```diff
@@ -722,6 +732,23 @@ DuckDbResultConvertor::fetchDataToArrowFormattedCiderBatch(
...
 }
 
+void updateChildrenNullCnt(struct ArrowArray* array) {
+  int64_t length = array->length;
+  for (int i = 0; i < array->n_children; i++) {
+    int null_count = 0;
+    auto child = array->children[i];
+    const uint8_t* validity_map = reinterpret_cast<const uint8_t*>(child->buffers[0]);
+    if (validity_map) {
+      for (int j = 0; j < length; j++) {
```
Comment (yma11) [inline src/cider/tests/utils/DuckDbQueryRunner.cpp:742]: Seems there is a method `CiderBitUtils::countSetBits` can be used.

**Comment by jikunshang** [comment]
@winningsix  @yma11 Please take a review, thanks.

File: src/cider/tests/utils/CiderNextgenQueryRunner.cpp [hunk 3]
```diff
@@ -58,7 +58,7 @@ std::string getFileContent(const std::string& file_name) {
 
 ::substrait::Plan CiderNextgenQueryRunner::genSubstraitPlan(
     const std::string& file_or_sql) {
-  INJECT_TIMER(GenSubstraitPlan);
+  // INJECT_TIMER(GenSubstraitPlan);
```
Comment (yma11) [inline src/cider/tests/utils/CiderNextgenQueryRunner.cpp:61]: remove directly?

File: src/cider/tests/nextgen/functional/benchmark/process_log.sh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+# cd $1
+
+files=`ls *INFO*log`
+# process each log file
+for file_name in $files
+do
+# file_name="NextgenBenchmarkTest.INFO.20221221-071202.log"
+
+# replace * with \*
+sed -i 's/*/\*/g' $file_name
```
Comment (yma11) [inline src/cider/tests/nextgen/functional/benchmark/process_log.sh:31]: Any introduction about how to use this script?

File: src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    auto duck_res_batch = DuckDbResultConvertor::fetchDataToCiderBatch(duck_res);
+  }
+  {
+    INJECT_TIMER(Nextgen);
+    struct ArrowArray output_array;
+    struct ArrowSchema output_schema;
+
+    auto file_or_sql = sql;
+    // auto file_or_sql = json_file.size() ? json_file : sql;
+    cider_nextgen_query_runner_->runQueryOneBatch(
```
Comment (yma11) [inline src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp:50]: How about run multi times?

File: src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+         CREATE_SUBSTRAIT_TYPE(I64),
+         CREATE_SUBSTRAIT_TYPE(Fp32),
+         CREATE_SUBSTRAIT_TYPE(Fp64)},
+        {},
+        GeneratePattern::Random,
+        -1000'000,
+        1000'000);
+  }
+};
+
```
Comment (kd97) [inline src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp:53]: need to provide some standard large-scaled dataset for benchmark test (eg move it to testbase.h)? btw, iter times can be set in different tests?

File: src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp (hunk 5 above)
Comment (jikunshang) [reply src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp:50]: I am not sure whether it's necessary to compare duck with cider run multi times. if it's necessary, I will extend NextegenRunner first.

File: src/cider/tests/nextgen/functional/benchmark/process_log.sh (hunk 4 above)
Comment (jikunshang) [reply src/cider/tests/nextgen/functional/benchmark/process_log.sh:31]: I will try to integrate this file into jenkins workflow later. this script will process log files which contains time breakdown and get a csv file for each log file.

File: src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp (hunk 6 above)
Comment (jikunshang) [reply src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp:53]: I could make this row number configurable in main function, please see latest update.
iteration time could only be one currently. I will try to support this if it's necessary.

**Review by yma11 — APPROVED** [review]

File: src/cider/tests/nextgen/functional/CMakeLists.txt [hunk 7]
```diff
@@ -30,3 +30,5 @@ foreach(next_gen_test_file ${next_gen_test_files})
   add_test(${target_name} ${EXECUTABLE_OUTPUT_PATH}/${target_name}
            --use_nextgen_compiler)
 endforeach()
+
+add_subdirectory(benchmark)
```
Comment (Ma-Jian1) [inline src/cider/tests/nextgen/functional/CMakeLists.txt:34]: move benchemark out of the test directory

File: src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp (hunk 5 above)
Comment (Ma-Jian1) [reply src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp:50]: small batch multi times may be better to tell the performance difference

File: src/cider/tests/utils/CiderNextgenTestBase.cpp [hunk 8]
```diff
@@ -39,12 +39,16 @@ void CiderNextgenTestBase::assertQuery(const std::string& sql,
...
   auto file_or_sql = json_file.size() ? json_file : sql;
-  cider_nextgen_query_runner_.runQueryOneBatch(
+  cider_nextgen_query_runner_->runQueryOneBatch(
       file_or_sql, *input_array_, *input_schema_, output_array, output_schema);
-  EXPECT_TRUE(CiderArrowChecker::checkArrowEq(duck_res_arrow[0].first.get(),
-                                              &output_array,
-                                              duck_res_arrow[0].second.get(),
-                                              &output_schema));
+  if (0 == duck_res_arrow.size()) {
+    // result is empty.
```
Comment (winningsix) [inline src/cider/tests/utils/CiderNextgenTestBase.cpp:45]: Should we check whether our output is empty?

File: src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+int main(int argc, char** argv) {
+  testing::InitGoogleTest(&argc, argv);
+
+  logger::LogOptions log_options(argv[0]);
+  log_options.parse_command_line(argc, argv);
+  logger::init(log_options);
+
+  // std::vector<int> row_nums{100'000, 1'000'000, 10'000'000};
```
Comment (winningsix) [inline src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp:73]: Remove this?

File: src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+#include "tests/utils/CiderNextgenBenchmarkBase.h"
+#include "tests/utils/CiderNextgenQueryRunner.h"
+#include "tests/utils/DuckDbQueryRunner.h"
+#include "tests/utils/QueryArrowDataGenerator.h"
+#include "util/measure.h"
+
+namespace cider::test::util {
+
+void CiderNextgenBenchmarkBase::benchSQL(const std::string& sql) {
```
Comment (winningsix) [inline src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp:34]: Hmm, why not use some benchmark framework like folly? It covers some basics like warming up, calculate some metrics like sec/op, performance relative ratio.

File: src/cider/tests/utils/CiderNextgenTestBase.cpp (hunk 8 above)
Comment (jikunshang) [reply src/cider/tests/utils/CiderNextgenTestBase.cpp:45]: I have another PR to fix this. blocked by this one.

File: src/cider/tests/utils/CiderArrowChecker.cpp [hunk 11]
```diff
@@ -138,16 +189,25 @@ bool checkOneScalarArrowEqual(const struct ArrowArray* expect_array,
     case 'l':
     case 'L':
       return checkArrowBuffer<int64_t>(expect_array, actual_array);
-    case 'e':
     case 'f':
+      return checkArrowBuffer<float>(expect_array, actual_array);
```
[the remark this answers is not in the document]
Comment (jikunshang) [reply src/cider/tests/utils/CiderArrowChecker.cpp:193]: will fix this in later PR.

File: src/cider/tests/nextgen/functional/CMakeLists.txt (hunk 7 above)
Comment (jikunshang) [reply src/cider/tests/nextgen/functional/CMakeLists.txt:34]: any suggestion location?

File: src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp (hunk 10 above)
Comment (jikunshang) [reply src/cider/tests/utils/CiderNextgenBenchmarkBase.cpp:34]: I think we don't need that in such scenario. We just need a time break down, right? it makes no sense to compare performance ratio between duckdb runner and cider runner.

File: src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+TEST_F(BenchmarkTest, singleFilter1) {
+  benchSQL("SELECT col_2 FROM test WHERE col_1 > 0");
+}
+
+int main(int argc, char** argv) {
+  testing::InitGoogleTest(&argc, argv);
+
+  logger::LogOptions log_options(argv[0]);
+  log_options.parse_command_line(argc, argv);
+  logger::init(log_options);
```
Comment (harborn) [inline src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp:71]: logger::LogOptions will remove, not use this again.

Comment (harborn) [reply src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp:71]: function logger::init also will remove.

Comment (jikunshang) [reply src/cider/tests/nextgen/functional/benchmark/NextgenBenchmarkTest.cpp:71]: so how to set log level in the future?

**Review by winningsix — APPROVED** [review]

**Review by yma11 — APPROVED** [review]

**Review by xieqi — APPROVED** [review]

**Pull request merged**